cmd/odo/odo.go

```diff
@@ -55,7 +55,7 @@ func main() {
 		return
 	}
 
-	cfg, err := preference.New()
+	cfg, err := preference.NewClient()
```

Comment:
How about replacing two occurrences of below in `odo.go`:
```go
util.LogErrorAndExit(root.ExecuteContext(segment.NewContext(context.Background())), "")
```
with below:
```go
util.LogErrorAndExit(root.ExecuteContext(segment.NewContext(context.WithValue(context.Background(), "preference", cfg))), "")
```
This way, every CLI's `Complete` function can access it with `cmdline.Context().Value("preference")`, and we can convert it to `preference.preferenceInfo` using something like below within the `preference` package:
```go
func GeneratePreferenceInfo(p interface{}) preferenceInfo {
	return p.(preferenceInfo)
}
```
WDYT? My motivation behind trying this is to reduce a bunch of calls to `preference.NewClient()` across the code to a single call. Sort of "don't repeat yourself" thing. It could also potentially help remove a bunch of `// TODO(feloy) Get from DI`, I guess.

Reply:
I would prefer not to use Context.Value, as it obscures the program's flow (see https://medium.com/@cep21/how-to-correctly-use-context-context-in-go-1-7-8f2c0fafdf39 - "Context.Value obscures your program’s flow" for detailed arguments)

Reply:
Thanks for sharing that article. It was a great read. Sorry about long reply. Correct me wherever you find the need to.

That article was heavily based on REST APIs, which I personally have very less experience with. I was trying to draw parallels with odo which is a CLI without a daemon process (loosely comparable to DB from that article).

I really loved the "inform vs. control" argument in that article.

In case of odo, if we do use `Context.Value`, I think, it should stay limited to the CLI layer. Layers below it - business logic & clients - should not use it directly in any way. Like the blog you shared suggests, we should create struct that is then passed to the below layers. Currently, below three calls to `preference.NewClient()` are making me cringe a bit since they are in layers below the CLI layer:
* `pkg/devfile/adapters/kubernetes/storage/utils.go`
* `pkg/kclient/oc_server.go`
* `pkg/kclient/pods.go`

I'm probably overthinking a condition which is making me cringe. I'll try to explain.

In case of `pkg/devfile/adapters/kubernetes/storage/utils.go`, the call is happening as a part of https://github.com/feloy/odo/blob/277a8c3f2b36b5a808bed9a0edbb88a60dd749bc/pkg/devfile/adapters/kubernetes/component/adapter.go#L199 which is after the call to https://github.com/feloy/odo/blob/277a8c3f2b36b5a808bed9a0edbb88a60dd749bc/pkg/devfile/adapters/kubernetes/component/adapter.go#L192.

Now, a hypothetical/crazy scenario I have thought of is what if `service.PushKubernetesResources` takes quite long (which is quite possible when creating a service) and user changes something in the preference before `a.createOrUpdateComponent` is called.

Irrespective of us using `Context.Value`, I feel that we should not be reading the preferences from the disk so deep into the code. It should be just passed along from CLI layer, and your approach of using Dependency Injection is absolutely cool.

> I would prefer not to use Context.Value

TBH, I have no preference. I just learned about `Context.Value` while doing that review 10 days back because I remembered reading about context but not being able to understand how it could be used. While doing the review, I found about `Context.WithValue` and tried it out. I'm just trying to have an open conversation and decide what's best.